File: src/main/java/com/github/tetrisanalyzer/piece/Piece.java
package com.github.tetrisanalyzer.piece;

/**
 * Represents one of the pieces: I, Z, S, J, L, T or O.
 */
public abstract class Piece {
    private static int[] ROTATION_MODULUS = { 0, 1, 0, 3 };
    private static Piece[] PIECES = { new PieceEmpty(), new PieceI(), new PieceZ(), new PieceS(), new PieceJ(), new PieceL(), new PieceT(), new PieceO(), new PieceAny() };

    public static int NUMBER_OF_PIECE_TYPES = 7;

    public static Piece get(int pieceNumber) {
        return PIECES[pieceNumber];
    }

    public abstract byte number();
    public abstract char character();
    public int rotationsEndIndex() { return heights().length - 1; }
    public int rotationModulus() { return ROTATION_MODULUS[rotationsEndIndex()]; }
    public int width(int rotation) { return widths()[rotation]; }
    public int height(int rotation) { return heights()[rotation]; }
    public PieceShape getShape(int rotation) { return shapes()[rotation]; }
    protected abstract int[] widths();
    protected abstract int[] heights();
    protected abstract PieceShape[] shapes();

    @Override
    public int hashCode() {
        return number();
    }

    @Override
    public boolean equals(Object obj) {
        if (!(obj instanceof Piece)) {
            return false;
        }
        return ((Piece)obj).number() == number();
    }

    @Override
    public String toString() {
        return String.valueOf(character());
    }
}

File: src/main/scala/com/github/tetrisanalyzer/piece/Piece.scala
package com.github.tetrisanalyzer.piece

object Piece {
  val NumberOfPieceTypes = 7
  private val RotationModulus: Array[Int] = Array(0, 1, 0, 3)
  private val Pieces: Array[Piece] = Array(
    new PieceEmpty, new PieceI, new PieceZ, new PieceS, new PieceJ, new PieceL, new PieceT, new PieceO, new PieceAny
  )
  private val Characters = Array.tabulate(NumberOfPieceTypes + 2) {
      ((i) => (Piece(i).character.toString))
  }

  private val Numbers = Map (
    Piece(0).character -> Piece(0).number,
    Piece(1).character -> Piece(1).number,
    Piece(2).character -> Piece(2).number,
    Piece(3).character -> Piece(3).number,
    Piece(4).character -> Piece(4).number,
    Piece(5).character -> Piece(5).number,
    Piece(6).character -> Piece(6).number,
    Piece(7).character -> Piece(7).number,
    Piece(8).character -> Piece(8).number
  )

  def apply(index: Int): Piece = Pieces(index)
  def toChar(number: Int) = Characters(number)
  def toNumber(character: Char) = Numbers(character)
}

/**
 * Represents one of the pieces: I, Z, S, J, L, T or O.
 */
abstract class Piece {
  def number: Byte
  def character: Char
  def rotationsEndIndex = heights.length - 1
  def rotationModulus = Piece.RotationModulus(rotationsEndIndex)
  def width(rotation: Int) = widths(rotation)
  def height(rotation: Int) = heights(rotation)
  def shape(rotation: Int) = shapes(rotation)
  protected def widths: Array[Int]
  protected def heights: Array[Int]
  protected def shapes: Array[PieceShape]

  override def hashCode = number

  override def equals(that: Any) = that match {
    case other: Piece => number == other.number
    case _ => false
  }

  override def toString = character.toString
}
